src/plugins/intel_cpu/src/graph.cpp

```diff
@@ -708,88 +705,282 @@ void Graph::ResolveComplexInplaceConflicts() {
+/**
+ * Partition the \clusters of Edges, by moving to the end and allocating at the same time
+ * the clusters which cannot be handled as part of generic memory solver algorithm.
+ * Such clusters meet one of the following criteria:
+ * - base edge of a cluster is already Allocated
+ * - base edge of a cluster is a "ov::element::string" type of edge
+ * - base edge of a cluster is a Constant edge
+ *
+ * @return a remaining number of clusters to process (left partition)
+ */
+static size_t AllocateStringsAndConstants(EdgeClusters& clusters,
+                                          const GraphContext::CPtr context) {
+    auto allocateConstantEdge = [context](const EdgePtr& edge) {
+        if (edge->getParent()->getType() == Type::Input) {
+            auto constNode = std::static_pointer_cast<node::Input>(edge->getParent());
+            edge->reuse(std::const_pointer_cast<IMemory>(constNode->getMemoryPtr()));
+        } else {
+            edge->externalAllocate(context->getWeightsCache());
+        }
+    };
+
+    auto allocateStringMemory = [context](const EdgePtr& edge) {
+        auto memory = std::make_shared<StringMemory>(context->getEngine(), edge->getDesc());
+        edge->reuse(memory);
+        return memory->getStringMemoryBlockPtr();
+    };
+
+    auto notAllocatedPartitionEnd =
+        std::partition(clusters.begin(), clusters.end(),
+                       [&allocateStringMemory, &allocateConstantEdge, &context](const EdgeCluster& cluster) {
+                           if (cluster.empty()) return false;
+
+                           auto baseEdgeIt = std::find_if(cluster.begin(), cluster.end(), [](const EdgePtr& edge) {
+                               return one_of(edge->getStatus(), Edge::Status::Allocated, Edge::Status::NeedAllocation);
+                           });
+
+                           OPENVINO_ASSERT(baseEdgeIt != cluster.end(), "Unexpected cluster state");
+
+                           const auto& baseEdge = *baseEdgeIt;
+                           if (baseEdge->getStatus() == Edge::Status::Allocated) {
+                               return false;
+                           }
+
+                           // Skip if the baseEdge does not require allocation
+                           if (baseEdge->getStatus() != Edge::Status::NeedAllocation) {
+                               return true;
+                           }
+
+                           // Allocate a cluster of the constants
+                           if (baseEdge->getParent()->isConstant()) {
+                               // @todo can we add some meaningful assert here?
+                               for (auto &edge : cluster) {
+                                   if (edge->getParent()->isConstant() && edge->getStatus() == Edge::Status::NeedAllocation) {
+                                       allocateConstantEdge(edge);
+                                   }
+                               }
+                               return false;
```

Comment:
To my understanding, only the base edge may have status `NeedAllocation`, and it's already know as the result of the search above. Do we really need to iterate once again?

Reply:
Is it true for the constants as well? I was trying to replicate the logic of the original loop, where possible, to not break anything.
Will try to remove it.

Reply:
@maxnick Can we apply it for the String logic as well?

Reply:
Yes, we can. The whole idea of the cluster is that it's a set of edges with the same memory block. And this memory block is located inside a specific memory object of a specific edge. So there can not be two edges with status `NeedAllocation` otherwise it's not clear what memory block should be shared across edges in the cluster.